common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,434 @@
+        handleOnDeviceCertAuth(request);
+    }
+
+    private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":handleSmartcardCertAuth";
+        //A connection to the YubiKey needs to be made in order to read the certificates off it.
+        mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
+                try {
+                    final SmartCardConnection c = value.getValue();
+                    PivSession piv = new PivSession(c);
+                    //Check if too many PIN attempts have been made.
+                    if (piv.getPinAttempts() == 0) {
+                        showMaxAttemptsDialog();
+                        request.cancel();
+                        Logger.infoPII(methodTag,  "User has reached the maximum failed attempts allowed.");
+                        return;
+                    }
+                    //Create List that contains cert details only pertinent to the cert picker.
+                    final List<YubiKitCertDetails> certList = getCertDetailsFromKey(piv);
+                    //If no certs were found, cancel flow.
+                    if (certList.isEmpty()) {
+                        request.cancel();
+                        Logger.infoPII(methodTag,  "No PIV certificates found on YubiKey device.");
+                        return;
+                    }
+                    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                    showSmartcardCertPickerDialog(certList, request);
+
+                } catch(IOException e) {
+                    Logger.errorPII(methodTag, "IOException", e);
+                    request.cancel();
+                } catch (ApduException e) {
+                    Logger.errorPII(methodTag, "ApduException", e);
+                    request.cancel();
+                } catch (ApplicationNotAvailableException e) {
+                    Logger.errorPII(methodTag, "ApplicationNotAvailableException", e);
+                    request.cancel();
+                }
+            }
+        });
+    }
+
+    //Helper method that returns a list of YubiKitCertDetails extracted from certificates located on the YubiKey.
+    //This method should only be called within a callback upon creating a successful YubiKey device connection.
+    private List<YubiKitCertDetails> getCertDetailsFromKey(@NonNull final PivSession piv) {
+        //Create ArrayList that contains cert details only pertinent to the cert picker.
+        List<YubiKitCertDetails> certList = new ArrayList<>();
+        //We need to check every slot. If a slot is empty, an ApduException is called, in which case, we'll ignore it.
```

Comment:
(If another type of exception is thrown - IOException/BadResponseException - should we still swallow it silently, or should we fail the whole operation?)

Reply:
TL;DR: any exception in `getAndPutCertDetailsInList` excluding FILE_NOT_FOUND ApduException will result in a cancelled flow and an error message. Altering both `getAndPutCertDetailsInLIst` and `GetCertDetailsFromKey` to throw the exceptions, and then `handleSmartcardAuth` can handle the exceptions and cancelling the flow.

Deleted the ignoring ApduException part of the comment, as it is already addressed in `getAndPutCertDetailsInList` and causes unnecessary confusion.
I looked into the `getCertificate` method of `PivSession`, and I think for IO and ApduException (excluding FILE_NOT_FOUND), we should definitely cancel the flow and show an error message.
I was going back and forth regarding BadResponseException, which seems to be thrown when the certificate is compressed or it cannot be parsed. On one hand, it wouldn't be great if the exception was thrown for a cert that would not be typically used for CBA, as the user could not authenticate even if their desired CBA cert was installed correctly. On the other hand, if the desired cert was the one throwing the BadResponseException, it would be super confusing to the user if they just didn't see that cert in the picker without any sort of error message. Since the latter scenario would probably be more likely to occur, I think the flow should be cancelled for BadResponseException too.
What I will do is alter both `getAndPutCertDetailsInLIst` and `GetCertDetailsFromKey` to throw the specified exceptions, and then `handleSmartcardAuth` can handle the exceptions and cancelling the flow.